Make Simulated DevTools Environment Panel Collapsible

victoreronmosele opened the pull request:
This PR updates the `SimulatedDevToolsWrapper` to make the environment panel collapsible. 

The panel's minimum size is removed and its width is now adjusted based on the available space.

Fixes #7021 

Before: 


https://github.com/flutter/devtools/assets/19398044/fbf5160a-e54b-43b9-9f03-a3d7070362bb



After: 



https://github.com/flutter/devtools/assets/19398044/74fcd996-cb95-4ea5-a5dd-be18845f02f1






## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or there is a reason for not adding tests.


![build.yaml badge]

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/devtools/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[build.yaml badge]: https://github.com/flutter/devtools/actions/workflows/build.yaml/badge.svg

kenzieschmoll commented on packages/devtools_extensions/integration_test/test/simulated_environment_test.dart:
```diff
@@ -192,6 +194,60 @@ Future<void> _testShowBannerMessage(
+      tester.renderObject(find.byWidget(environmentPanel));
+  final double environmentPanelRenderBoxWidth =
+      environmentPanelRenderBox.size.width;
+
```
at this point, can you add an `expect` to check that the width is what we'd expect? The value of `_VmServiceConnection._totalControlsWidth + 2 * defaultSpacing`?

victoreronmosele replied:
Thanks for the review. 

I added this but it's failing with this message:

```
Expected: <439>
Actual: <435.7075>
```

Looking into why.

victoreronmosele replied:
The width from the `RenderBox` wasn't the expected width but close (not sure why).  
So in https://github.com/flutter/devtools/pull/7062/commits/9c6a468e4b84784d3208ec20818bd071c14d5811, I updated the tests to get the width from the `SizedBox` wrapping the expansion panel.

kenzieschmoll commented on packages/devtools_extensions/lib/src/template/_simulated_devtools_environment/_simulated_devtools_environment.dart:
```diff
@@ -74,73 +75,107 @@ class SimulatedDevToolsWrapperState extends State<SimulatedDevToolsWrapper>
+                  availableEnvironmentPanelWidth,
+                );
+
+                return SingleChildScrollView(
```
can you add a Scrollbar to this so that it is clear this section is horizontally scrollable?

victoreronmosele replied:
I've added this now. Please see [screen recording](https://github.com/flutter/devtools/pull/7062#issuecomment-1905315146).

kenzieschmoll reviewed (commented):
Thanks for the contribution. Just a couple requests.

kenzieschmoll reviewed (commented):
Can you please also add an entry to `devtools_extensions/CHANGELOG.md` and update the `detools_app/release_notes/NEXT_RELEASE_NOTES.md` file with an entry under the "## DevTools Extension updates" section. Thanks!

victoreronmosele commented:
Here's how the environment panel looks with the `Scrollbar`:

https://github.com/flutter/devtools/assets/19398044/3d36d029-34c6-4cb0-90e0-31c4a630dd8e

kenzieschmoll commented on packages/devtools_extensions/lib/src/template/_simulated_devtools_environment/_connect_ui.dart:
```diff
@@ -4,22 +4,24 @@
 
-  static const _totalControlsHeight = 45.0;
-  static const _totalControlsWidth = 415.0;
+  static const totalControlsHeight = 45.0;
+  static const totalControlsWidth = 415.0;
```
nit: mark these `@visibleForTesting`

victoreronmosele replied:
Done.

kenzieschmoll reviewed (approved):
one nit, then LGTM! Also, the failing test needs to be fixed.

victoreronmosele commented:
@kenzieschmoll I've fixed the failing test. 

Previously, I used the `Rect.overlaps()` method to check if the environment panel is collapsed, but this failed due to a rounding issue — environmentPanelRect.left was returning `1599.9999999999998` instead of `1600` as expected. 

To fix this issue, I updated the overlap check to be done directly in the test file, rounding the `environmentPanelRect.left` value up to the nearest integer with `.ceil()`.
